lib/honeybadger/events_worker.rb

```diff
@@ -0,0 +1,329 @@
+      @pid = Process.pid
+      @send_queue = []
+      @last_sent = nil
+
+      @max_events = config.get(:events_batch_size)
+      @send_timeout = config.get(:events_timeout)
+    end
+
+    def push(msg)
+      return false unless start
+
+      if queue.size >= config.max_queue_size
+        warn { sprintf('Unable to send event; reached max queue size of %s.', queue.size) }
+        return false
+      end
+
+      queue.push(msg)
+    end
+
+    def send_now(msg)
+      handle_response(send_to_backend(msg))
+    end
+
+    def shutdown(force = false)
+      d { 'shutting down worker' }
+
+      mutex.synchronize do
+        @shutdown = true
+      end
+
+      return true if force
+      return true unless thread&.alive?
+
+      if throttled?
+        warn { sprintf('Unable to send %s event(s) to Honeybadger (currently throttled)', queue.size) } unless queue.empty?
+        return true
+      end
+
+      info { sprintf('Waiting to send %s events(s) to Honeybadger', queue.size) } unless queue.empty?
+      queue.push(FLUSH)
+      queue.push(SHUTDOWN)
+      !!thread.join
+    ensure
+      queue.clear
+      kill!
+    end
+
+    # Blocks until queue is processed up to this point in time.
+    def flush
+      mutex.synchronize do
+        if thread && thread.alive?
+          queue.push(FLUSH)
+          queue.push(marker)
+          marker.wait(mutex)
+        end
+      end
+    end
+
+    def start
+      return false unless can_start?
+
+      mutex.synchronize do
+        @shutdown = false
+        @start_at = nil
+
+        return true if thread&.alive?
+
+        @pid = Process.pid
+        @thread = Thread.new { run }
+        @timeout_thread = Thread.new { schedule_timeout_check }
+      end
+
+      true
+    end
+
+    private
+
+    attr_reader :config, :queue, :pid, :mutex, :marker, :thread, :timeout_thread, :throttle,
+      :throttle_interval, :start_at, :send_queue, :last_sent, :max_events, :send_timeout
+
+    def_delegator :config, :backend
+
+    def shutdown?
+      mutex.synchronize { @shutdown }
+    end
+
+    def suspended?
+      mutex.synchronize { start_at && Time.now.to_i < start_at }
+    end
+
+    def can_start?
+      return false if shutdown?
+      return false if suspended?
+      true
+    end
+
+    def throttled?
+      mutex.synchronize { throttle > 0 }
+    end
+
+    def kill!
+      d { 'killing worker thread' }
+
+      if thread
+        Thread.kill(thread)
+        Thread.kill(timeout_thread)
+        thread.join # Allow ensure blocks to execute.
+      end
+
+      true
+    end
+
+    def suspend(interval)
+      mutex.synchronize do
+        @start_at = Time.now.to_i + interval
+        queue.clear
+      end
+
+      # Must be performed last since this may kill the current thread.
+      kill!
+    end
+
+    def schedule_timeout_check
+      loop do
+        sleep(send_timeout / 1000.0)
+        ms_since = (Time.now.to_f - last_sent.to_f) * 1000.0
+        if ms_since >= send_timeout
```

Comment:
This code appears to be duplicated with `#check_timeout`. Do both methods need to perform the send_timout check? If so, you could add a predicate helper method that both methods could share.

Reply:
I'll opt for the latter, as I think there's good chance a timeout check could be scheduled and then send based on queue size could happen before the actual `timeout_check` method is called.

Reply:
(On second thought, maybe simply throwing the CHECK_TIMEOUT message into the queue at regular intervals should be enough)